addons/binding/org.openhab.binding.cm11a/src/main/java/org/openhab/binding/cm11a/internal/X10Interface.java

```diff
@@ -0,0 +1,800 @@
+                nextModule = deviceUpdateQueue.take();
+                log.trace("Got a device.  Going to run it.");
+
+                // Keep retrying to update this device until it is successful.
+                boolean success = false;
+                while (!success) {
+                    try {
+                        if (connect()) {
+                            nextModule.updateHardware(this);
+                            success = true;
+                        } else {
+                            Thread.sleep(IO_RECONNECT_INTERVAL);
+                        }
+                    } catch (IOException e) {
+                        connected = false;
+                        log.error("IO Exception when updating module hardware.  Will retry shortly", e);
+                        Thread.sleep(IO_RECONNECT_INTERVAL);
+                    } catch (InvalidAddressException e) {
+                        log.error("Attempted to send an X10 Function call with invalid address.  Ignoring this.");
+                        success = true; // Pretend this was successful as retrying will be pointless.
+                    }
+                }
+
+            } catch (InterruptedException e1) {
+                log.warn("Unexpected interrupt on X10 scheduling thread.  Ignoring and continuing anyway...");
+            }
+        }
+        log.trace("Stopping background thread...");
+        this.notifyAll();
+    }
+
+    @Override
+    public void serialEvent(SerialPortEvent event) {
+        try {
+            if (event.getEventType() == SerialPortEvent.DATA_AVAILABLE || event.getEventType() == SerialPortEvent.RI) {
+                synchronized (serialPort) {
+
+                    log.trace("Serial port data available or RI indicator event received");
+                    while (serialPort.isRI() && serialInput.available() <= 0) {
+                        log.trace("Ring indicator is High but there is no data.  Waiting for data...");
+                        try {
+                            Thread.sleep(20);
+                        } catch (InterruptedException e) {
+                            log.warn("Interrupted while sleeping", e);
+                        }
+                    }
+
+                    while (serialInput.available() > 0) {
+                        log.trace(serialInput.available() + " bytes of data available to be read");
+                        int readint = serialInput.read();
+
+                        processRequestFromIFace(readint);
+
+                        // Wait a while before rechecking to give interface time to switch off Ring Indicator.
+                        while (serialPort.isRI() && serialInput.available() <= 0) {
+                            log.trace("Ring indicator is High but there is no data.  Waiting for data...");
+                            try {
+                                Thread.sleep(300);
+                            } catch (InterruptedException e) {
```

Comment:
If we receive this InterruptedException, the interrupted flag was cleared and we most likely need to reset the flag:

http://www.yegor256.com/2015/10/20/interrupted-exception.html
https://www.ibm.com/developerworks/library/j-jtp05236/